### PR TITLE
Campaign gamemode

### PR DESCRIPTION
**Pull request by Lumipharon**
## About The Pull Request
Still heavily WIP. Message me if you want to help with mapping or spriting.

Adds a new gamemode. Might be terrible.

I'll explain in some more detail but here's a brief rundown for now.

### **About the game mode**
Campaign is an attempt to significantly depart from our normal gameplay loop.
Instead of a shipside and groundside, campaign revolves around two factions having multiple battles against each other, across and around a planet in an attempt to wrestle control of them planet for themselves.

### Missions
Gameplay is divided into a series of missions, each with its own set of objects, particular scenario, and rewards.
When a mission is loaded, **a new z-level is loaded specifically for that mission**. So as a campaign progresses, players will fight across a whole range of maps, trying to achieve a variety of goals to ultimately drive the other faction of out of the system.

Deployments is typically via tad rappel like existing HvH modes, but TGMC have access to drop pods (including mech and sentry pods) while SOM have an AOE bluespace teleporter.
image
image

The faction that wins a mission then has the opportunity to select the next mission that is played, choosing from a number of options they have available to them.

image

### Victory Points
Victory Points (VP) are accrued by winning missions, which either allow the faction to play a 'end mission' to conclusively win the campaign, or otherwise grind down the enemy team by scoring a VP (minor) victory.

While VP is the ultimate points of the gamemode, it is not the only thing played for during missions.

### Campaign Assets
Both factions have access to a variety of campaign assets to help them in their fight. These can be won as mission rewards, or purchased with attrition points (AP).

Assets can range from gear upgrades, deployment options, allied support to air support.

image
image

image

Some "assets" can be negative, and are usually assigned to the **opposing** team as a reward for certain mission which can do things like reduce your AP generation rate, to permanently disabling certain assets you would normally have access to.

### Attrition Points
While attrition points can be spent on assets, but their primary purpose is to allow for player respawns.
Every mission, your faction leader determines how many attrition points will be spent on respawns. Unspent points are lost however, so simply spamming points can be wasteful, and a team on the backfoot could even deliberately avoid spending points to save them up for a cheeky counter attack in a later mission, etc.

### Code stuff
Thanks to @TiviPlus for the tgui work, which I am utterly useless at.

Codewise, Campaign is not hardcoded to specific factions like existing modes are, so changing the factions (to sectoids vs CLF, marines vs xenos, etc) is extremely easy.

Missions being plug and play also means it would be simple to create future campaigns to represent other planets or stories, so there's some coolness that could be had down the line.

Most of everything is done, but a lot more work is needed to get all the mission I want in play to a ready state. In particular, a lot of mapping is required, so any mappers out there, feel free to reach out if you want to lend a hand.

There's also a variety of art assets I want or need, so ditto on spriters.

I want to get campaign into a testable state relatively quickly however, to get an idea of how this actually ends up playing, and how to tweak it going forwards (or if its even good at all).
## Why It's Good For The Game
Is it?
## Changelog
:cl:
add: Added Campaign gamemode
/:cl:

### CODE REVIEW

**Comment by Araulius** [comment]
will this be taking inspiration from red orchestra's campaign mode? I'm curious to see how this goes

**Comment by Oscari-pog** [comment]
This sounds cool

File: code/__DEFINES/jobs.dm [hunk 1]
```diff
@@ -63,10 +63,23 @@
 #define SQUAD_CORPSMAN "Squad Corpsman"
 #define SQUAD_ENGINEER "Squad Engineer"
 #define SQUAD_MARINE "Squad Marine"
+#define SQUAD_ROBOT "Squad Robot"
 #define SQUAD_VATGROWN "Squad VatGrown"
 #define SILICON_AI "AI"
 
-//SOM - very placeholder for now
+//SOM
+#define SOM_GENERAL "SOM general"
```
Comment (XSlayer300) [inline code/__DEFINES/jobs.dm:71]:
```suggestion
#define SOM_GENERAL "SOM General"
```

File: code/__DEFINES/jobs.dm [hunk 2]
```diff
@@ -63,10 +63,23 @@
...
+#define SOM_FIELD_COMMANDER "SOM Field Commander"
+#define SOM_STAFF_OFFICER "SOM Staff Officer"
+#define SOM_PILOT_OFFICER "SOM Pilot Officer"
+#define SOM_MECH_PILOT "SOM Mech Pilot"
+#define SOM_REQUISITIONS_OFFICER "SOM Requisitions Officer"
+#define SOM_CHIEF_ENGINEER "SOM Chief Engineer"
+#define SOM_CHIEF_MEDICAL_OFFICER "SOM Chief Medical Officer"
+#define SOM_TECH "SOM Technician"
+#define SOM_MEDICAL_DOCTOR "SOM Medical Doctor"
+#define SOM_CHEF "SOM chef"
```
Comment (XSlayer300) [inline code/__DEFINES/jobs.dm:81]:
```suggestion
#define SOM_CHEF "SOM Chef"
```

File: code/datums/fire_support.dm [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+	uses = -1
+
+/datum/fire_support/incendiary_rockets
+	name = "Incendiary rocket barrage"
+	fire_support_type = FIRESUPPORT_TYPE_INCEND_ROCKETS
+	scatter_range = 9
+	impact_quantity = 12
+	icon_state = "incendiary_rockets"
+	initiate_chat_message = "TARGET ACQUIRED ROCKET RUN INBOUND."
+	initiate_screen_message = "Rockets hot, incoming!"
```
Comment (XSlayer300) [inline code/datums/fire_support.dm:154]:
```suggestion
	initiate_screen_message = "Incendiary missiles away, stay off the heat!"
```

File: code/datums/gamemodes/campaign/campaign_reward_jobs.dm [hunk 4]
```diff
@@ -0,0 +1,624 @@
...
+	H.equip_to_slot_or_del(new /obj/item/ammo_magazine/rifle/mpi_km/black, SLOT_IN_BELT)
+	H.equip_to_slot_or_del(new /obj/item/ammo_magazine/rifle/mpi_km/black, SLOT_IN_BELT)
+	H.equip_to_slot_or_del(new /obj/item/ammo_magazine/rifle/mpi_km/black, SLOT_IN_BELT)
+	H.equip_to_slot_or_del(new /obj/item/ammo_magazine/rifle/mpi_km/black, SLOT_IN_BELT)
+
+	H.equip_to_slot_or_del(new /obj/item/explosive/grenade/som, SLOT_IN_BACKPACK)
+	H.equip_to_slot_or_del(new /obj/item/explosive/grenade/som, SLOT_IN_BACKPACK)
+
+/datum/job/som/mercenary/militia/medic
+	title = "Militial Medic"
```
Comment (XSlayer300) [inline code/datums/gamemodes/campaign/campaign_reward_jobs.dm:190]:
```suggestion
	title = "Militia Medic"
```

File: code/datums/gamemodes/campaign/campaign_rewards.dm [hunk 5]
```diff
@@ -0,0 +1,517 @@
...
+		if(teleporter.faction != faction.faction)
+			continue
+		teleporter.teleporter_status = TELEPORTER_ARRAY_INOPERABLE
+		to_chat(faction.faction_leader, span_warning("Error: The Teleporter Array has been rendered permanently inoperable."))
+		return
+
+/datum/campaign_reward/teleporter_charges
+	name = "Delegated Teleporter Array access"
+	desc = "+2 uses of the Teleporter Array"
+	detailed_desc = "Central command have allocated the battalion with two additional uses of the Teleporter Array. Its extremely costly to run and demand is high across the conflict zone, so make them count."
```
Comment (XSlayer300) [inline code/datums/gamemodes/campaign/campaign_rewards.dm:262]:
```suggestion
	detailed_desc = "Central Command have allocated the battalion with two additional uses of the Teleporter Array. It's extremely costly to run and demand is high across the conflict zone, so make them count."
```

File: code/datums/gamemodes/campaign/missions/fire_support_raid.dm [hunk 6]
```diff
@@ -0,0 +1,62 @@
+//disabling some of the enemy's firesupport options
+/datum/campaign_mission/destroy_mission/fire_support_raid
+	name = "Fire support raid"
+	map_name = "Lunar base BD-832"
```
Comment (XSlayer300) [inline code/datums/gamemodes/campaign/missions/fire_support_raid.dm:4]:
```suggestion
	map_name = "Lunar Base BD-832"
```

File: code/datums/gamemodes/campaign/missions/loot_capture.dm [hunk 7]
```diff
@@ -0,0 +1,144 @@
+//Loot capture mission
+/datum/campaign_mission/capture_mission
+	name = "Capture mission"
+	map_name = "Jungle outpost SR-422"
```
Comment (XSlayer300) [inline code/datums/gamemodes/campaign/missions/loot_capture.dm:4]:
```suggestion
	map_name = "Jungle Outpost SR-422"
```

File: code/datums/gamemodes/campaign/missions/teleporter_raid.dm [hunk 8]
```diff
@@ -0,0 +1,62 @@
+//disabling SOM's ability to teleport deploy
+/datum/campaign_mission/destroy_mission/teleporter_raid
+	name = "Teleporter control raid"
+	map_name = "Lunar base BD-832"
```
Comment (XSlayer300) [inline code/datums/gamemodes/campaign/missions/teleporter_raid.dm:4]:
```suggestion
	map_name = "Lunar Base BD-832"
```

File: code/datums/jobs/job/sons_of_mars_shipside.dm [hunk 9]
```diff
@@ -0,0 +1,758 @@
...
+		<b>You answer to</b> SOM High Command<br /><br />
+		<b>Unlock Requirement</b>: Starting Role<br /><br />
+		<b>Gamemode Availability</b>: Campaign<br /><br /><br />
+		<b>Duty</b>: Lead the SOM forces and complete your mission. Support the marines and communicate with your command staff, execute orders.
+	"}
+	minimap_icon = "captain" //placeholder
+
+/datum/job/som/command/general/radio_help_message(mob/M)
+	. = ..()
+	to_chat(M, {"As the captain of the [SSmapping.configs[SHIP_MAP].map_name] you are held by higher standard and are expected to act competently.
```
Comment (XSlayer300) [inline code/datums/jobs/job/sons_of_mars_shipside.dm:39]:
```suggestion
	to_chat(M, {"As the field general of the [SSmapping.configs[SHIP_MAP].map_name] you are held by higher standard and are expected to act competently.
```

File: code/datums/jobs/job/sons_of_mars_shipside.dm [hunk 10]
```diff
@@ -0,0 +1,758 @@
...
+While you may support Nanotrasen, you report to the SOM High Command, not the corporate office.
```
Comment (XSlayer300) [inline code/datums/jobs/job/sons_of_mars_shipside.dm:40]: either remove this or replace it with something else

File: code/datums/jobs/job/sons_of_mars_shipside.dm [hunk 11]
```diff
@@ -0,0 +1,758 @@
...
+Your primary task is the safety of the ship and her crew, and ensuring the survival and success of the marines.
+Your first order of business should be briefing the marines on the mission they are about to undertake.
```
Comment (XSlayer300) [inline code/datums/jobs/job/sons_of_mars_shipside.dm:42]:
```suggestion
Your first order of business should be briefing the soldiers on the mission they are about to undertake.
```

File: code/datums/jobs/job/sons_of_mars_shipside.dm [hunk 12]
```diff
@@ -0,0 +1,758 @@
...
+		<b>Gamemode Availability</b>: Campaign<br /><br /><br />
+		<b>Duty</b>: Lead the SOM forces and complete your mission. Support the marines and communicate with your command staff, execute orders.
+	"}
+	minimap_icon = "captain" //placeholder
+
+/datum/job/som/command/general/radio_help_message(mob/M)
+	. = ..()
+	to_chat(M, {"As the captain of the [SSmapping.configs[SHIP_MAP].map_name] you are held by higher standard and are expected to act competently.
+While you may support Nanotrasen, you report to the SOM High Command, not the corporate office.
+Your primary task is the safety of the ship and her crew, and ensuring the survival and success of the marines.
```
Comment (XSlayer300) [inline code/datums/jobs/job/sons_of_mars_shipside.dm:41]:
```suggestion
Your primary task is the safety of the ship and her crew, and ensuring the survival and success of the Sons of Mars.
```

File: code/game/objects/structures/campaign_structure.dm [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+	pixel_x = -16
+	mission_types = list(/datum/campaign_mission/destroy_mission/teleporter_raid)
+	objective_type = /obj/structure/campaign_objective/destruction_objective/bluespace_core
+
+#define BLUESPACE_CORE_OK "bluespace_core_ok"
+#define BLUESPACE_CORE_UNSTABLE "bluespace_core_unstable"
+#define BLUESPACE_CORE_BROKEN "bluespace_core_broken"
+
+/obj/structure/campaign_objective/destruction_objective/bluespace_core
+	name = "\improper Bluespace Teleportation Displacement Core"
```
Comment (XSlayer300) [inline code/game/objects/structures/campaign_structure.dm:78]:
```suggestion
	name = "\improper Bluespace teleportation displacement core"
```

**Comment by blackdav123** [comment]
whats the intended playtime of this mode and the individual missions? is this supposed to be a 3 hour affair with 45-60 minute missions or a ~90 minute gamemode with 20 minute missions?

**Comment by Lumipharon** [comment]
Missions will likely be 15-30 minutes (at a guess) but the whole campaign could go on for hours.
Because each mission is like a small game in of itself, the overall length doesn't worry me too much since people can just dip out after a mission ends or whatever

**Comment by Mannybrado** [comment]
"Dang, this look so good!"
>and SOM...
Oh, it's HvH... 💀

**Comment by Oscari-pog** [comment]
Will you be able to keep captured gear? Like, if you manage to snag a B18 off a dead marine, and SoM wins the mission, will you get to take the B18 back to the SoM ship or whatever? I still dont entirely understand some of the ideas, but this sounds absolutely based.

**Comment by Lumipharon** [comment]
Yes. Anyone alive at the end of a mission is beamed back to base. You loot it, you keep it.

However some asset stuff is specifically for one mission only. I.e. you call in CAS and its for that mission, any unused uses are wasted.
Same with mechs.

**Comment by iforgotmeword** [comment]
how would med-evacs work?
and would you be able to sell enemy command roles like civil war?

**Comment by Aporhtonoma** [comment]
is this hvh only intented or will xenos be included in the form of xeno boiler cas bombardments and assets like that?

**Comment by bowlofcereal** [comment]
planetside 2

**Comment by Mannybrado** [comment]
this would seriously go so hard if it had a xeno version. so hard. 
but you... youre soft... arent you...

**Review by Kuro020 — APPROVED** [review]
Pre approval of sprites so I dont need to keep checking the PR

**Comment by Kuro020** [comment]
> this would seriously go so hard if it had a xeno version. so hard. but you... youre soft... arent you...

"Codewise, Campaign is not hardcoded to specific factions like existing modes are, so changing the factions (to sectoids vs CLF, marines vs xenos, etc) is extremely easy."

**Comment by lbnesquik** [comment]
I'll take my name in the credits of this PR thank you very much

**Comment by Joe13413** [comment]
holy based, this sounds fun

**Comment by pilotmans** [comment]
Singlehandedly making HvH fun

**Review by spookydonut — APPROVED** [review]
Let's save lumi from merge conflict hell and fix any remaining issues with follow up PRs

**Pull request merged**